### PR TITLE
Unify menus cleanups

### PR DESCRIPTION
**Pull request by jelly**
Merge the two menu's into one helper function, this reduces a ton of code duplication, simplifies and unifies everything. As side-effect removes renaming the current directory.

### CODE REVIEW

File: src/fileActions.jsx [hunk 1]
```diff
@@ -563,3 +520,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+
+    const createLink = (currentPath, files, selected) => {
+        Dialogs.show(
+            <CreateLinkModal
+              currentPath={currentPath} selected={selected}
+              files={files.sort((a, b) => (a.name.toLowerCase() > b.name.toLowerCase())
+                  ? 1
+                  : ((b.name.toLowerCase() > a.name.toLowerCase())
+                      ? -1
+                      : 0))}
```
Comment (jelly) [inline src/fileActions.jsx:538]: This is some unreadable stuff..

Comment (cockpituous) [inline src/fileActions.jsx:538]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240306-175329-4e4bd6c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 2]
```diff
@@ -563,3 +520,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+            { type: "divider" },
+            {
+                id: "create-item",
+                title: _("Create directory"),
+                onClick: () => Dialogs.show(<CreateDirectoryModal currentPath={currentPath} />),
+            },
+            {
+                id: "create-link",
+                title: _("Create link"),
+                onClick: () => createLink(currentPath, files, selected || {})
```
Comment (cockpituous) [inline src/fileActions.jsx:566]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240306-175329-4e4bd6c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 3]
```diff
@@ -563,3 +520,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+            },
+            { type: "divider" },
+            {
+                id: "edit-permissions",
+                title: _("Edit permissions"),
+                onClick: () => editPermissions(Dialogs, selected[0] || { name: directory_name, type: "dir" }, path)
```
Comment (cockpituous) [inline src/fileActions.jsx:572]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240306-175329-4e4bd6c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 4]
```diff
@@ -563,3 +520,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+                        onClick: () => spawnPaste(clipboard, [currentPath + selected[0].name], false, addAlert),
+                        isDisabled: clipboard === undefined
+                    }
+                ]
+                : [],
+            { type: "divider" },
+            {
+                id: "edit-permissions",
+                title: _("Edit permissions"),
+                onClick: () => editPermissions(Dialogs, selected[0] || { name: directory_name, type: "dir" }, path)
```
Comment (cockpituous) [inline src/fileActions.jsx:597]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240306-175329-4e4bd6c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 5]
```diff
@@ -563,3 +520,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+            },
+            {
+                id: "rename-item",
+                title: _("Rename"),
+                onClick: () => {
+                    Dialogs.show(
+                        <RenameItemModal
+                          path={path}
+                          selected={selected[0] || { name: directory_name, type: "dir" }}
```
Comment (cockpituous) [inline src/fileActions.jsx:606]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240306-175329-4e4bd6c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 6]
```diff
@@ -563,3 +520,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+                    );
+                }
+            },
+        );
+    } else if (selected.length > 1) {
+        menuItems.push(
+            {
+                id: "copy-item",
+                title: _("Copy"),
+                onClick: () => setClipboard(selected.map(s => path.join("/") + "/" + s.name)),
```
Comment (cockpituous) [inline src/fileActions.jsx:637]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240306-175329-4e4bd6c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 7]
```diff
@@ -394,6 +351,11 @@ export const EditPermissionsModal = ({ selected, path }) => {
     const groups = useFile("/etc/group", { syntax: etcGroupSyntax });
     const logindef = useFile("/etc/login.defs", { superuser: true });
 
+    if (!selected) {
+        const directory_name = path[path.length - 1];
+        selected = { name: directory_name, type: "dir" };
```
Comment (cockpituous) [inline src/fileActions.jsx:356]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-084539-34707c5b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 8]
```diff
@@ -563,3 +526,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+
+    const createLink = (currentPath, files, selected) => {
+        Dialogs.show(
+            <CreateLinkModal
+              currentPath={currentPath} selected={selected}
+              files={files.sort((a, b) => (a.name.toLowerCase() > b.name.toLowerCase())
+                  ? 1
+                  : ((b.name.toLowerCase() > a.name.toLowerCase())
+                      ? -1
+                      : 0))}
```
Comment (cockpituous) [inline src/fileActions.jsx:544]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-084539-34707c5b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 9]
```diff
@@ -563,3 +526,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+            { type: "divider" },
+            {
+                id: "create-item",
+                title: _("Create directory"),
+                onClick: () => Dialogs.show(<CreateDirectoryModal currentPath={currentPath} />),
+            },
+            {
+                id: "create-link",
+                title: _("Create link"),
+                onClick: () => createLink(currentPath, files, selected || {})
```
Comment (cockpituous) [inline src/fileActions.jsx:572]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-084539-34707c5b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 10]
```diff
@@ -563,3 +526,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+            },
+            { type: "divider" },
+            {
+                id: "edit-permissions",
+                title: _("Edit permissions"),
+                onClick: () => editPermissions(Dialogs, selected[0], path)
```
Comment (cockpituous) [inline src/fileActions.jsx:578]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-084539-34707c5b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 11]
```diff
@@ -563,3 +526,137 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+                    );
+                }
+            },
+        );
+    } else if (selected.length > 1) {
+        menuItems.push(
+            {
+                id: "copy-item",
+                title: _("Copy"),
+                onClick: () => setClipboard(selected.map(s => path.join("/") + "/" + s.name)),
```
Comment (cockpituous) [inline src/fileActions.jsx:643]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-084539-34707c5b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx (hunk 7 above)
Comment (cockpituous) [inline src/fileActions.jsx:356]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-085857-be9ac358-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 12]
```diff
@@ -563,3 +525,135 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+
+    const createLink = (currentPath, files, selected) => {
+        Dialogs.show(
+            <CreateLinkModal
+              currentPath={currentPath} selected={selected}
+              files={files.sort((a, b) => (a.name.toLowerCase() > b.name.toLowerCase())
+                  ? 1
+                  : ((b.name.toLowerCase() > a.name.toLowerCase())
+                      ? -1
+                      : 0))}
```
Comment (cockpituous) [inline src/fileActions.jsx:541]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-085857-be9ac358-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 13]
```diff
@@ -563,3 +525,135 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+            { type: "divider" },
+            {
+                id: "create-item",
+                title: _("Create directory"),
+                onClick: () => Dialogs.show(<CreateDirectoryModal currentPath={currentPath} />),
+            },
+            {
+                id: "create-link",
+                title: _("Create link"),
+                onClick: () => createLink(currentPath, files, selected || {})
```
Comment (cockpituous) [inline src/fileActions.jsx:569]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-085857-be9ac358-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 14]
```diff
@@ -563,3 +525,135 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+            },
+            { type: "divider" },
+            {
+                id: "edit-permissions",
+                title: _("Edit permissions"),
+                onClick: () => editPermissions(Dialogs, selected[0], path)
```
Comment (cockpituous) [inline src/fileActions.jsx:575]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-085857-be9ac358-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 15]
```diff
@@ -563,3 +525,135 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+                    );
+                }
+            },
+        );
+    } else if (selected.length > 1) {
+        menuItems.push(
+            {
+                id: "copy-item",
+                title: _("Copy"),
+                onClick: () => setClipboard(selected.map(s => path.join("/") + "/" + s.name)),
```
Comment (cockpituous) [inline src/fileActions.jsx:640]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-085857-be9ac358-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 16]
```diff
@@ -563,3 +525,135 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+                title: _("Edit permissions"),
+                onClick: () => editPermissions(Dialogs, selected[0], path)
+            }
+        );
+    } else if (selected.length === 1) {
+        menuItems.push(
+            {
+                id: "copy-item",
+                title: _("Copy"),
+                // HACK: maybe it is the same as below?
```
Comment (jelly) [inline src/fileActions.jsx:583]: Drop the hack.

File: src/fileActions.jsx (hunk 7 above)
Comment (cockpituous) [inline src/fileActions.jsx:356]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-100054-4e7bfb3b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 17]
```diff
@@ -563,3 +525,134 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+
+    const createLink = (currentPath, files, selected) => {
+        Dialogs.show(
+            <CreateLinkModal
+              currentPath={currentPath} selected={selected}
+              files={files.sort((a, b) => (a.name.toLowerCase() > b.name.toLowerCase())
+                  ? 1
+                  : ((b.name.toLowerCase() > a.name.toLowerCase())
+                      ? -1
+                      : 0))}
```
Comment (cockpituous) [inline src/fileActions.jsx:541]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-100054-4e7bfb3b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 18]
```diff
@@ -563,3 +525,134 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+            { type: "divider" },
+            {
+                id: "create-item",
+                title: _("Create directory"),
+                onClick: () => Dialogs.show(<CreateDirectoryModal currentPath={currentPath} />),
+            },
+            {
+                id: "create-link",
+                title: _("Create link"),
+                onClick: () => createLink(currentPath, files, selected || {})
```
Comment (cockpituous) [inline src/fileActions.jsx:569]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-100054-4e7bfb3b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 19]
```diff
@@ -563,3 +525,134 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+            },
+            { type: "divider" },
+            {
+                id: "edit-permissions",
+                title: _("Edit permissions"),
+                onClick: () => editPermissions(Dialogs, selected[0], path)
```
Comment (cockpituous) [inline src/fileActions.jsx:575]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-100054-4e7bfb3b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 20]
```diff
@@ -563,3 +525,134 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+                    );
+                }
+            },
+        );
+    } else if (selected.length > 1) {
+        menuItems.push(
+            {
+                id: "copy-item",
+                title: _("Copy"),
+                onClick: () => setClipboard(selected.map(s => path.join("/") + "/" + s.name)),
```
Comment (cockpituous) [inline src/fileActions.jsx:639]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-100054-4e7bfb3b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx (hunk 7 above)
Comment (cockpituous) [inline src/fileActions.jsx:356]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-112122-d90b1ce5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx (hunk 17 above)
Comment (cockpituous) [inline src/fileActions.jsx:541]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-112122-d90b1ce5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx (hunk 18 above)
Comment (cockpituous) [inline src/fileActions.jsx:569]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-112122-d90b1ce5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx (hunk 19 above)
Comment (cockpituous) [inline src/fileActions.jsx:575]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-112122-d90b1ce5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx [hunk 21]
```diff
@@ -563,3 +525,134 @@ const setDirectoryName = (val, setName, setNameError, setErrorMessage) => {
...
+                          selected={selected[0]}
+                        />
+                    );
+                },
+            },
+            { type: "divider" },
+            {
+                id: "create-link",
+                title: _("Create link"),
+                onClick: () => createLink(currentPath, files, selected[0]),
```
Comment (cockpituous) [inline src/fileActions.jsx:617]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-112122-d90b1ce5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/fileActions.jsx (hunk 20 above)
Comment (cockpituous) [inline src/fileActions.jsx:639]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-112122-d90b1ce5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/navigator-card-body.jsx [hunk 22]
```diff
@@ -345,6 +271,8 @@ export const NavigatorCardBody = ({
             <>
                 {contextMenu}
                 <ListingTable
+                  onClick={resetSelected}
+                  onContextMenu={resetSelected}
```
Comment (cockpituous) [inline src/navigator-card-body.jsx:275]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-112122-d90b1ce5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/navigator-card-body.jsx [hunk 23]
```diff
@@ -374,6 +301,7 @@ export const NavigatorCardBody = ({
 const Item = React.memo(function Item({ file, isSelected, setSelected, onDoubleClickNavigate, isGrid }) {
     function handleClick(ev, file) {
         if (ev.detail > 1) {
+            setSelected([]);
```
Comment (cockpituous) [inline src/navigator-card-body.jsx:304]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-307-20240307-112122-d90b1ce5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
Thanks! How difficult is it to [avoid the extra space](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1485-34d6b3b7-20240308-095846-fedora-39/log.html)? If "not easy", then I'm happy to just accept the new references.

File: src/fileActions.jsx (hunk 1 above)
Comment (martinpitt) [reply src/fileActions.jsx:541]: It's okay-ish code-wise, but it doesn't make sense to me here. First of all, you can probably shorten this to `a.toLowerCase().localeCompare(b.toLowerCase())

But moreover, this seems like the wrong place. The "files in current directory" state of the main application should already be sorted (I assume). So why do you need to re-sort here? The dialog should show the same sorting as the main view.

This is just moved code, but cleaning this up would be a good follow-up.

**Review by martinpitt — APPROVED** [review]
Thanks for the nice cleanups! One follow-up remark, but that doesn't need to be a blocker.
